benchmark/projection/include/create_data.hpp

```diff
@@ -0,0 +1,168 @@
+  return npar;
+}
+
+static inline std::vector<int> simplesol_dist(int ncell, int active_cells,
+                                              int min_per_cell,
+                                              int max_per_cell) {
+  if (ncell == active_cells)
+    return dense_dist(ncell, min_per_cell, min_per_cell);
+  if (ncell < active_cells) {
+    fprintf(stderr, "Active cells > total cells\n");
+    exit(1);
+  }
+  // std::random_device device;
+  assert(min_per_cell > 0);
+  assert(max_per_cell >= min_per_cell);
+  std::mt19937 random(2983); // device());
+  std::uniform_int_distribution<> pdist(min_per_cell, max_per_cell);
+  std::uniform_int_distribution<> cdist(0, ncell);
+  std::vector<int> npar(ncell, 0);
+
+  // anticipating ncell >> active_cells so this should be ok
+  while (active_cells > 0) {
+    int cell = cdist(random);
+    if (npar[cell] == 0) {
+      active_cells--;
+      npar[cell] = pdist(random);
+    }
+  }
+  return npar;
+}
+
+template <int nmode, typename T, typename Shape>
+static inline auto create_data(sycl::queue &Q, int ncell, int min_per_cell,
+                               int max_per_cell, int active_cells = -1) {
+  // Shove all the pointers in a vector so we can free them later
+
+  std::vector<void *> all_pointers;
+
+  // Par per cell
+  int *par_per_cell = sycl::malloc_device<int>(ncell, Q);
+  assert(par_per_cell);
+  all_pointers.push_back((void *)par_per_cell);
+  int max_row = 0;
+  std::vector<int> host_par_per_cell;
+  if (active_cells < 0 || active_cells == ncell) {
+    host_par_per_cell = dense_dist(ncell, min_per_cell, max_per_cell);
+    max_row =
+        *std::max_element(host_par_per_cell.begin(), host_par_per_cell.end());
+    Q.copy<int>(host_par_per_cell.data(), par_per_cell, ncell).wait();
+  } else {
+    host_par_per_cell =
+        simplesol_dist(ncell, active_cells, min_per_cell, max_per_cell);
+    max_row =
+        *std::max_element(host_par_per_cell.begin(), host_par_per_cell.end());
+    Q.copy<int>(host_par_per_cell.data(), par_per_cell, ncell).wait();
+  }
+  // DOFS
+  T *dofs =
+      sycl::malloc_device<T>(ncell * Shape::template get_ndof<nmode>(), Q);
+  assert(dofs);
+  all_pointers.push_back((void *)dofs);
+  Q.fill(dofs, T{0.0}, ncell * Shape::template get_ndof<nmode>())
+      .wait_and_throw();
+
+  // DOF offsets ndof per cell but scanned
+  auto host_offsets =
+      std::vector<int>(ncell, Shape::template get_ndof<nmode>());
+  std::exclusive_scan(host_offsets.begin(), host_offsets.end(),
+                      host_offsets.begin(), 0);
+  int *dof_offsets = sycl::malloc_device<int>(ncell, Q);
+  assert(dof_offsets);
+  all_pointers.push_back((void *)dof_offsets);
+  Q.copy<int>(host_offsets.data(), dof_offsets, ncell).wait_and_throw();
+
+  // Cell ids
+  int *cell_ids = sycl::malloc_device<int>(ncell, Q);
+  assert(cell_ids);
+  all_pointers.push_back((void *)cell_ids);
+  Q.parallel_for<>(ncell, [=](sycl::id<1> id) { cell_ids[id] = id; }).wait();
```

Comment:
I cannot see a `parallel_for` in the standard that takes a size_t/int as the first arg?
```suggestion
  Q.parallel_for<>(sycl::range<1>(ncell), [=](sycl::id<1> id) { cell_ids[id] = id; }).wait();
```

Reply:
it's implicit construction of the range from the int - but you're right it's better to explicit